### PR TITLE
cpu/lpc2387: update the UART driver

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

The UART driver on lpc2387 was hard-coding one UART with a set baud rate.

This updates the driver to 

 - allow the board to configure the RX & TX pins using a config struct like other UART drivers
 - allow for more than one UART
 - allow setting the baudrate
 - implement poweron()/poweroff() functions

### Testing procedure

I tested the two UARTs on the MCB2388 board using 9600/57600/115200 Baud.

Please verify that the UART on the `msba2` still works.

I also added a config for the remaining UARTs on the msba2 which are exposed through JP2, JP3 & JP4 according to the schematics.
I don't have one of those boards to test the configuration, but it should be pretty straight forward.

### Issues/PRs references
Partially addresses #4693

### CODE REVIEW

**Comment by maribu** [comment]
Very cool! Thanks! I'll test this afternoon.

File: cpu/lpc2387/periph/uart.c [hunk 1]
```diff
@@ -29,84 +30,251 @@
...
+#endif
+#if UART_NUMOF > 3
+    UART3_IRQHandler,
+#endif
+};
+
+/* Table for FDR register contents
+ * bits 0…3: MULVAL
+ * bits 4…7: DIVADDVAL
+ */
```
Comment (leandrolanzieri) [inline cpu/lpc2387/periph/uart.c]: I think this may be the other way around? [p.428](https://www.nxp.com/docs/en/user-guide/UM10211.pdf)

Comment (leandrolanzieri) [reply cpu/lpc2387/periph/uart.c]: Although the table seems to be ok, I think it's only a matter of the parameters name

File: cpu/lpc2387/include/periph_cpu.h [hunk 2]
```diff
@@ -19,15 +19,13 @@
...
-#include "cpu.h"
-#include "periph/dev_enums.h"
-
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 #include <stdint.h>
 #include "cpu.h"
+#include "kernel_defines.h"
```
Comment (leandrolanzieri) [inline cpu/lpc2387/include/periph_cpu.h]: Why is this needed?

File: boards/msba2/include/periph_conf.h [hunk 3]
```diff
@@ -87,7 +87,31 @@ static const uart_conf_t uart_config[] = {
         .pinsel_tx   = 0,
         .pinsel_msk_rx = BIT4,
         .pinsel_msk_tx = BIT6,
-    }
+    },
+    {
+        .dev = UART1,
+        .irq_prio_rx = 6,
```
Comment (leandrolanzieri) [inline boards/msba2/include/periph_conf.h]: Is there any logic or considerations needed for the IRQ priority level on these?

File: boards/msba2/include/periph_conf.h [hunk 4]
```diff
@@ -79,8 +79,42 @@ extern "C" {
...
+        .dev = UART3,
+        .irq_prio_rx = 6,
+        .pinsel_rx   = 9,
+        .pinsel_tx   = 9,
+        .pinsel_msk_rx = BIT26 | BIT27,
+        .pinsel_msk_tx = BIT28 | BIT29,
+    },
+};
+
 #define UART_NUMOF          (1)
```
Comment (leandrolanzieri) [inline boards/msba2/include/periph_conf.h]:
```suggestion
#define UART_NUMOF          (4)
```

File: boards/msba2/include/periph_conf.h [hunk 5]
```diff
@@ -79,8 +79,42 @@ extern "C" {
...
+        .pinsel_msk_rx = BIT22,
+        .pinsel_msk_tx = BIT20,
+    },
+    {
+        .dev = UART3,
+        .irq_prio_rx = 6,
+        .pinsel_rx   = 9,
+        .pinsel_tx   = 9,
+        .pinsel_msk_rx = BIT26 | BIT27,
+        .pinsel_msk_tx = BIT28 | BIT29,
```
Comment (leandrolanzieri) [inline boards/msba2/include/periph_conf.h]: I think this should be
```suggestion
        .pinsel_msk_tx = BIT24 | BIT25,
```

Comment (benpicco) [reply boards/msba2/include/periph_conf.h]: Good catch!

File: cpu/lpc2387/include/periph_cpu.h (hunk 2 above)
Comment (benpicco) [reply cpu/lpc2387/include/periph_cpu.h]: That was for `ARRAY_SIZE()` but since that didn't work with preprocessor arithmetic comparisons, I had to remove it again.

File: cpu/lpc2387/periph/uart.c (hunk 1 above)
Comment (benpicco) [reply cpu/lpc2387/periph/uart.c]: Oh yes - I had it reversed at first and then forgot to update the comment.

File: boards/msba2/include/periph_conf.h (hunk 3 above)
Comment (benpicco) [reply boards/msba2/include/periph_conf.h]: Since the VIC supports 32 interrupts with 16 levels, two (or more) interrupts at the same level should be no problem.

I used 6 because that's what has been used before.

**Comment by leandrolanzieri** [comment]
I tested it on the `msba2`, on all 4 UARTs using different baudrates and everything looks perfect:

**Frame checking**:
image

<details><summary><b>Echo test on UART 3</b></summary>

```
> init 3 115200
2019-11-04 14:24:06,827 #  init 3 115200
2019-11-04 14:24:06,829 # Success: Initialized UART_DEV(3) at BAUD 115200
2019-11-04 14:24:07,084 # UARD_DEV(3): test uart_poweron() and uart_poweroff()  ->  [OK]
> send 3 RIOT!
2019-11-04 14:24:12,085 #  send 3 RIOT!
2019-11-04 14:24:12,086 # UART_DEV(3) TX: RIOT!
> 2019-11-04 14:24:12,101 #  Success: UART_DEV(3) RX: [RIOT!]\n
```
</details>

File: cpu/lpc2387/periph/uart.c [hunk 6]
```diff
@@ -29,84 +30,251 @@
...
+#if UART_NUMOF > 3
+    UART3_IRQHandler,
+#endif
+};
+
+/* Table for FDR register contents
+ * bits 0…3: DIVADDVAL
+ * bits 4…7: MULVAL
+ */
+#define DIV_MUL(m, d) ((d << 4) | m)
```
Comment (leandrolanzieri) [inline cpu/lpc2387/periph/uart.c]: Still, this looks a bit misleading, as `m` divides and `d` multiplies.

File: cpu/lpc2387/periph/uart.c [hunk 7]
```diff
@@ -29,84 +30,251 @@
...
+static const uint8_t div_table[] = {
+    DIV_MUL(0,  1),  /* 1       */
+    DIV_MUL(0,  1),  /* 1.03125 */
```
Comment (leandrolanzieri) [inline cpu/lpc2387/periph/uart.c]: Repeated?

File: cpu/lpc2387/periph/uart.c [hunk 8]
```diff
@@ -29,84 +30,251 @@
...
+    DIV_MUL(2,  3),  /* 1.6875  */
+    DIV_MUL(5,  7),  /* 1.71875 */
+    DIV_MUL(3,  4),  /* 1.75    */
+    DIV_MUL(7,  9),  /* 1.78125 */
+    DIV_MUL(9, 11),  /* 1.8125  */
+    DIV_MUL(11,13),  /* 1.84375 */
+    DIV_MUL(7,  8),  /* 1.875   */
+    DIV_MUL(9, 13),  /* 1.90625 */
+    DIV_MUL(14,15),  /* 1.9375  */
+    DIV_MUL(14,15),  /* 1.96875 */
```
Comment (leandrolanzieri) [inline cpu/lpc2387/periph/uart.c]: Repeated?

File: cpu/lpc2387/periph/uart.c (hunk 7 above)
Comment (benpicco) [reply cpu/lpc2387/periph/uart.c]: The next smallest possible value here is `1 + 1/15` or 1.0625.
Since `1 + 1/32` is closer to 1 than to 1.0625, I decided to round down.

File: cpu/lpc2387/periph/uart.c (hunk 6 above)
Comment (benpicco) [reply cpu/lpc2387/periph/uart.c]: fixed.
